File: backend/src/modules/users/infrastructure/persistence/dynamo-user.repository.ts
import {
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
  QueryCommand,
} from '@aws-sdk/lib-dynamodb';
import { User } from '@modules/users/domain/entities/user.entity';
import { UserRepository } from '@modules/users/domain/repositories/user.repository';
import {
  KEY_PREFIXES,
  SK_VALUES,
  TABLE,
} from '@shared/infrastructure/dynamodb/single-table.constants';
import { UserMapper } from './user.mapper';

export class DynamoUserRepository implements UserRepository {
  constructor(
    private readonly client: DynamoDBDocumentClient,
    private readonly tableName: string,
  ) {}

  async findByEmail(email: string): Promise<User | null> {
    const result = await this.client.send(
      new QueryCommand({
        TableName: this.tableName,
        IndexName: TABLE.GSI1,
        KeyConditionExpression: 'GSI1PK = :pk AND GSI1SK = :sk',
        ExpressionAttributeValues: {
          ':pk': `${KEY_PREFIXES.EMAIL}${email.toLowerCase()}`,
          ':sk': SK_VALUES.USER,
        },
        Limit: 1,
      }),
    );

    if (!result.Items || result.Items.length === 0) return null;
    return UserMapper.toDomain(result.Items[0]);
  }

  async findById(id: string): Promise<User | null> {
    const result = await this.client.send(
      new GetCommand({
        TableName: this.tableName,
        Key: {
          PK: `${KEY_PREFIXES.USER}${id}`,
          SK: SK_VALUES.PROFILE,
        },
      }),
    );

    if (!result.Item) return null;
    return UserMapper.toDomain(result.Item);
  }

  async save(user: User): Promise<void> {
    await this.client.send(
      new PutCommand({
        TableName: this.tableName,
        Item: UserMapper.toItem(user),
        ConditionExpression: 'attribute_not_exists(PK)',
      }),
    );
  }
}

File: backend-v2/crates/infrastructure/src/dynamodb/user/dynamo_user_repository.rs
use async_trait::async_trait;
use aws_sdk_dynamodb::operation::put_item::PutItemError;
use aws_sdk_dynamodb::types::AttributeValue;
use aws_sdk_dynamodb::Client;
use domain::modules::user::entity::User;
use domain::modules::user::repository::{RepositoryError, UserRepository};

use crate::dynamodb::user::user_mapper::UserMapper;

pub struct DynamoUserRepository {
    client: Client,
    table_name: String,
}

impl DynamoUserRepository {
    pub fn new(client: Client, table_name: String) -> Self {
        Self { client, table_name }
    }
}

#[async_trait]
impl UserRepository for DynamoUserRepository {
    async fn find_by_email(&self, email: &str) -> Result<Option<User>, RepositoryError> {
        let output = self
            .client
            .query()
            .table_name(&self.table_name)
            .index_name("GSI1")
            .key_condition_expression("GSI1PK = :pk")
            .expression_attribute_values(":pk", AttributeValue::S(email.to_string().to_lowercase()))
            .limit(1)
            .send()
            .await
            .map_err(|e| RepositoryError::DatabaseError(e.to_string()))?;

        let items = output.items.unwrap_or_default();
        match items.into_iter().next() {
            Some(item) => UserMapper::to_domain(&item).map(Some),
            None => Ok(None),
        }
    }

    async fn save(&self, user: &User) -> Result<(), RepositoryError> {
        self.client
            .put_item()
            .table_name(&self.table_name)
            .set_item(Some(UserMapper::to_item(user)))
            .condition_expression("attibute_not_exists(PK)")
            .send()
            .await
            .map_err(|err| match err.into_service_error() {
                PutItemError::ConditionalCheckFailedException(_) => RepositoryError::AlreadyExists,
                other => RepositoryError::DatabaseError(other.to_string()),
            })?;

        Ok(())
    }
}
